### DISCUSSION TITLE
Return to state you came from

### DISCUSSION THREAD

**Discussion by lanmonster** [statelyai/xstate#1788] [Q&A] [answered]
I have a situation that I don't really know how to solve with this library.

3 states: A, B, C

A -> C is valid
B -> C is valid

C -> A is valid if and only if A -> C happened most recently
C -> B is valid if and only if B -> C happened most recently

I have tried using history state but I was running into the issue of not being able to know 1 state deep in the stack (you can only know the most recent, or top of the stack).

Any thoughts?
Thanks!

**Comment by Andarist** [comment]
You could assign a variable to context when exiting A and B states and just make guarded transitions from C based on that.

**Comment by davidkpiano** [comment, marked as answer]
Like @Andarist said, you can use guards for this. You can also think of your problem differently: instead of saying "A -> C" then "C -> A only if A -> C happened", translate that last part into an actual _state_:

> A -> C_from_A is valid
> B -> C_from_B is valid

Then, you can make use of compound states:

```js
states: {
  A: {
    on: { EVENT: 'C.A' }
  },
  B: {
    on: { EVENT: 'C.B' }
  },
  C: {
    initial: 'A', // doesn't matter
    states: {
      A: { on: { BACK: 'A' } },
      B: { on: { BACK: 'B' } }
    }
  }
});
```

**Reply by Andarist** [reply]
@davidkpiano neat approach with compound state 👍

**Reply by lanmonster** [reply]
What complicates it a bit is that A-> B and B -> A is also valid. I ended up with this machine and I think I can work with this solution
```
Machine({
    id: 'test',
    initial: 'A',
    states: {
      A: {
        on: { EVENT: 'C.from_a', B: 'B' }
      },
      B: {
        on: { EVENT: 'C.from_b', A: 'A' }
      },
      C: {
        initial: 'from_a',
        states: {
          from_a: { on: { BACK: '#test.A' } },
          from_b: { on: { BACK: '#test.B' } }
        }
      }
    }
  });
```

Thanks for your swift responses!

**Reply by f-elix** [reply]
I think an alternative solution would be to wrap states A and B in a parent state AB. A can go to B and vice versa, and both can go to C. You can then use a history state on state AB for when you enter it again from the C state. I have not tested it but it seems to me to be the solution with the least number of states.
